### ISSUE TITLE
import photos without defined identifiers and create the link to a context later

### ISSUE THREAD

**Issue by lbestock** [urapadmin/kiosk#694]
We have discussed this and I know it's big and I cannot remember the totality of our conversation. It came up again today in talking to Candace and Tyler as a desirable workflow - they wanted to be able to do this and expected it - and so I wanted to make sure it had a ticket - I couldn't find one.

Currently there has to be a record for a photo to be attached to already for an imported photo to end up having an archaeological context. Particularly for legacy projects where there are records and there are photographs but not yet any link between them, and where the photographs are not labeled or sorted in a way that makes their context clear, it would be very helpful to be able to do a bulk unsorted photo import and do the assigning in kiosk itself. 

Related and less critical but potentially worth discussing is if contexts themselves could be created by giving an identifier to a previously unidentified photo in kiosk. As in, I am looking at an image, see a wall that was not previously given a locus number, and add a locus identification to that photograph that then creates that locus. Sounds abstract and ridiculous but it more or less happened the other day when I was looking at images of FA and describing things and wham, there's a wall that no one ever called anything. I went back into the recording system, made the locus, went to the gallery, and found images to put in that way. It wasn't too bad, I could use tags in the gallery to help, but it would have been easier from the images themselves.

**Comment by urapadmin** [comment]
as you said, these are two different things. 
a) assigning an image to an existing context -> yes, that is possible and not such a big deal since we have the context management in kiosk. BUT: The recording data has to be there first, because, and here comes b:
b) creating a context identifier out of an image is not feasible. You would have to state all kinds of information, e.G. what kind of unit? What kind of locus? Identification methods would have to decide what the identifier has to look like etc. etc. The recording system does not only record a bunch of identifiers but also a lot of data. I see no reasonable way to do this nor do I have the feeling it is the right approach. If we were able to type in recording data in the browser, that is just what you would have done. Before we don't get there, this can't happen. We must go with the split of inputing data in filemaker and I do not expect that to change before my term at Brown ends, frankly. I would be glad if we could at least look at data in the browser. Even that is far away.

**Comment by urapadmin** [comment]
so here we focus on assigning an image to an existing context. The first, easier attempt, will do that in the file edit dialog. But I do see it coming that people want to bulk assign images soon. That will be harder since we do not have any bulk operations at all so far (see #20 (my god, 20! That ticket is ages old).

**Comment by urapadmin** [comment]
### kiosk 0.8.7
was also not such a small deal since what they want to add they might also want to delete (and if it is only because a formerly added context was wrong). This needs intensive testing with non-critical data in a test project before we can roll it out to real projects. Particularly USTP with their hacky collected materials that do not have identifiers 

- [ ] simple: Adding a identifier should attach the image to the record type that is responsible for storing files for that identifier by default. I know, sounds wrought: A unit identifier stores an image in dayplans, a locus identifier in locus_photo (which can be called something else depending on the project) usw. We cannot assign an image to anything but the default record type. e.g. we cannot assign an image to a locus relation of locus "FH-001", simply because we cannot address let alone create the particular relation.
- [ ] simple: detaching the image again from the identifier (there is a little red button next to any single identifier) will just break the connection between the record in the default file record type and the image, The record will be empty. e.g. detaching an image from locus_photo will leave the locus_photo record with an empty image. Why don't we just delete the locus_photo record? Because it might very well have other data that we don't want to lose. At least all those records can have an image description. 
- [ ] more complicated: What about detaching images from record types that are not the default record types. e.G. detach an image from a locus relation.

**Comment by urapadmin** [comment]
test failed: detaching an image from locus relations does not work because locus_relation images are not default file locations and cannot be, currently. I have to distinguish between the default file location which I need when adding an image and all file locations that can be connected to an identifier. That way I should also be able to solve the USTP issue with the collected material that does not have its own identifier.

**Comment by urapadmin** [comment]
see #725 and #728 where I stumbled over the destination for adding a file to a context as well.

**Comment by urapadmin** [comment]
### kiosk 0.8.7 again

- [ ] simple: Adding a identifier should attach the image to the record type that is responsible for storing files for that identifier by default. I know, sounds wrought: A unit identifier stores an image in dayplans, a locus identifier in locus_photo (which can be called something else depending on the project) usw. We cannot assign an image to anything but the default record type. e.g. we cannot assign an image to a locus relation of locus "FH-001", simply because we cannot address let alone create the particular relation.
- [ ] simple: detaching the image again from the identifier (there is a little red button next to any single identifier) will just break the connection between the record in the default file record type and the image, The record will be empty. e.g. detaching an image from locus_photo will leave the locus_photo record with an empty image. Why don't we just delete the locus_photo record? Because it might very well have other data that we don't want to lose. At least all those records can have an image description. 
- [ ] more complicated: What about detaching images from record types that are not the default record types. e.G. detach an image from a locus relation.
- [ ] weird things happen: If you have a collected material that has a uronarti and a wheeler id (so it has two context identifiers) both of them are shown in the edit file dialog. No matter which one you delete, none of them will reappear after deleting one of them: Beause the image is not assigned any longer to the collected material. That is correct, though admittedly befuddling.
- [ ] weirder: if you have an image in  locus relations both sides of the relation appear in the list of identifiers. If you delete only one of them BOTH will remain in the list. WHAT? But yes. Although one of the relations lost the image indeed, the relation still connects it with the reverse relation and there the image is still there. That is, well, befuddling, too, but it is technically correct. Only if you delete both contexts from the list, both disappear. I accept this as proper behavior.

**Comment by urapadmin** [comment]
something that is missing: When a relation to a file gets added (e.g. a locus_photo relation) that relation does not get the proper logged on user assigned to it. Instead it gets "sys". Not a disaster but I should attend to it one day.

**Comment by lbestock** [comment]
This ticket got baroque. I have tested all sorts of things around this and think it is working properly. If I add context descriptions and break the context I maintain the empty record and full description. I do note that the locus relations is weird. One cannot add that relation in as one can add other relations - one would have to upload the image, give it a context so it appears in the gallery, and then insert it in FM to the relationship. And breaking is weird. If one has a relationship between LA-001 and LA-002 and one goes in and deletes LA-001, as you said I should expect it doesn't really go away, I open the image again and now it says LA-002 and LA-001 (having reversed the order, now recreating LA-001 from LA-002). BUT if I then break the link to LA-002, but not in that case LA-001, it indeed loses both of them. So it recreated it, but only sort of - not enough to keep the whole relationship from being deleted when I deleted just the one.

As for the alternate number issue, I have not touched that as I'm testing in PVD. If you want me to I will reconfigure things.

**Comment by lbestock** [comment]
You have attended to the sys issue in the reciprocal locus relation, which had become its own ticket and was closed. I think this can be closed.

But the thing with alternate numbers is still weird. I will describe what I did and you tell me if this is what is expected; if so close this.

I created a CM with a photo in the recording system (10.3.6 on my iPad). It had its automatically assigned PVD number and I also gave it an alternate number, Wheeler 28.1. I gave it a description.

I uploaded and synchronized and all was as expected.

I clicked the file in the file repository, and asked it to break the Wheeler number. It broke that connection, as well as the connection to the PVD CM#. It also deleted the description, which I think was proper since that would have been the contextual description; the only information on this photo now is the date.

But the ticket was to test applying and creating links to photos that come in without context. So I took an imported context-less photo and assigned it two contexts, one a unit and one the Wheeler 28.1. I was a bit surprised to see that the latter worked - the number got in the system. And was connected with the PVD# I made it in, as that got assigned as a context automatically by the system once my alternate number was in there. So two contexts were applied though I added only one. I added a context independent description. I then broke the alternate number. As before, this also broke the principal CM number. The context-independent description was maintained, as expected.
